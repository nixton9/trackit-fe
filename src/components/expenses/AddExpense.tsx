import React, { useState } from 'react'
import DatePickerInput from '../misc/DatePickerInput'
import { Styled } from '../../styles/Add.styles'
import { FluidInput } from '../misc/FluidInput'
import { AddSubmitButton } from '../misc/Add'
import { CustomAddSelect } from '../misc/CustomAddSelect'
import { CustomAddDatePicker } from '../misc/CustomAddDatePicker'
import { ReactComponent as NotesIcon } from '../../assets/icons/notes.svg'
import { expensesCategories } from '../../assets/fakeData'
import { ExpenseCategory } from '../../utils/ModuleTypes'

const AddExpense: React.FC = () => {
  const [value, setValue] = useState<number | string | undefined>(undefined)
  const [title, setTitle] = useState('')
  const [category, setCategory] = useState('0')
  const [date, setDate] = useState(new Date())

  const categoryOptions = [
    { val: '0', label: 'All' },
    ...(expensesCategories as ExpenseCategory[]).map(cat => ({
      val: cat.id,
      label: cat.name
    }))
  ]

  const handleCategoryChange = (e: React.ChangeEvent<HTMLSelectElement>) => {
    setCategory(e.target.value)
  }

  const handleSubmit = () => {
    console.log(value)
    console.log(title)
    console.log(category)
    console.log(date)
  }

  return (
    <>
      <Styled.AddInputNumberWrapper>
        <FluidInput value={value} setValue={setValue} placeholder={'9.99'} />
        <span className={value ? 'active' : ''}>â‚¬</span>
      </Styled.AddInputNumberWrapper>

      <Styled.AddInput
        type="text"
        placeholder="Ex: Dinner at mcdonalds"
        value={title}
        onChange={e => setTitle(e.target.value)}
      />

      <Styled.AddWidgetsContainer>
        <Styled.AddWidget>
          <DatePickerInput
            date={date}
            setDate={setDate}
            customInput={<CustomAddDatePicker />}
          />
        </Styled.AddWidget>

        <Styled.AddWidget>
          <CustomAddSelect
            id="add-category"
            value={category}
            onChange={handleCategoryChange}
            options={categoryOptions}
            icon={<NotesIcon />}
          />
        </Styled.AddWidget>

        <AddSubmitButton handleSubmit={handleSubmit} />
      </Styled.AddWidgetsContainer>
    </>
  )
}

export default AddExpense
