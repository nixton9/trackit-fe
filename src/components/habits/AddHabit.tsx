import React, { useState } from 'react'
import { AddSubmitButton } from '../misc/Add'
import { Styled } from '../../styles/Add.styles'

const AddHabit: React.FC = () => {
  const [title, setTitle] = useState('')

  const handleSubmit = () => {
    console.log(title)
  }

  return (
    <>
      <Styled.AddInput
        type="text"
        placeholder="Ex: Eat healthy"
        value={title}
        onChange={e => setTitle(e.target.value)}
      />

      <Styled.AddWidgetsContainer>
        <AddSubmitButton handleSubmit={handleSubmit} />
      </Styled.AddWidgetsContainer>
    </>
  )
}

export default AddHabit
