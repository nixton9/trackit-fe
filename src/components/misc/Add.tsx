import React, { useState, useEffect } from 'react'
import Drawer from './Drawer'
import AddNote from '../notes/AddNote'
import AddTask from '../tasks/AddTask'
import AddHabit from '../habits/AddHabit'
import AddExpense from '../expenses/AddExpense'
import { useToggleElement } from '../../utils/useToggleElement'
import { ModuleTypes } from '../../utils/ModuleTypes'
import { ReactComponent as PlusIcon } from '../../assets/icons/plus.svg'
import { ReactComponent as NotesIcon } from '../../assets/icons/notes.svg'
import { ReactComponent as TasksIcon } from '../../assets/icons/tasks.svg'
import { ReactComponent as ExpensesIcon } from '../../assets/icons/expenses.svg'
import { ReactComponent as HabitsIcon } from '../../assets/icons/habits.svg'
import { ReactComponent as ChevronIcon } from '../../assets/icons/chevron.svg'
import { Styled } from '../../styles/Add.styles'

const notesTitle = 'Create a note'
const tasksTitle = 'Create a task'
const expensesTitle = 'Add an expense'
const habitsTitle = 'Add an habit'

export const AddSubmitButton: React.FC<{ handleSubmit: () => void }> = ({
  handleSubmit
}) => (
  <Styled.AddWidget__Button onClick={handleSubmit}>
    <ChevronIcon />
  </Styled.AddWidget__Button>
)

const Add: React.FC = () => {
  const [activeContent, setActiveContent] = useState<ModuleTypes | null>(null)
  const [open, setOpen, overlayEl] = useToggleElement()

  useEffect(() => {
    if (open) setActiveContent(null)
  }, [open])

  let drawerTitle
  let drawerContent

  switch (activeContent) {
    case ModuleTypes.Notes:
      drawerTitle = notesTitle
      drawerContent = <AddNote closeModal={() => setOpen(false)} />
      break

    case ModuleTypes.Tasks:
      drawerTitle = tasksTitle
      drawerContent = <AddTask />
      break

    case ModuleTypes.Expenses:
      drawerTitle = expensesTitle
      drawerContent = <AddExpense />
      break

    case ModuleTypes.Habits:
      drawerTitle = habitsTitle
      drawerContent = <AddHabit />
      break

    default:
      drawerTitle = 'What do you want to track?'
      drawerContent = (
        <>
          <Styled.AddButton onClick={() => setActiveContent(ModuleTypes.Notes)}>
            <span>{notesTitle}</span>
            <NotesIcon />
          </Styled.AddButton>

          <Styled.AddButton onClick={() => setActiveContent(ModuleTypes.Tasks)}>
            <span>{tasksTitle}</span>
            <TasksIcon />
          </Styled.AddButton>

          <Styled.AddButton
            onClick={() => setActiveContent(ModuleTypes.Expenses)}
          >
            <span>{expensesTitle}</span>
            <ExpensesIcon />
          </Styled.AddButton>

          <Styled.AddButton
            onClick={() => setActiveContent(ModuleTypes.Habits)}
          >
            <span>{habitsTitle}</span>
            <HabitsIcon />
          </Styled.AddButton>
        </>
      )
  }

  return (
    <>
      <Styled.AddIcon onClick={() => setOpen(true)}>
        <PlusIcon />
      </Styled.AddIcon>

      <Drawer title={drawerTitle} open={open} overlayRef={overlayEl}>
        {drawerContent}
      </Drawer>
    </>
  )
}

export default Add
