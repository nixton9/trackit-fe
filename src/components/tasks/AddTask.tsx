import React, { useState } from 'react'
import DatePickerInput from '../misc/DatePickerInput'
import { Styled } from '../../styles/Add.styles'
import { AddSubmitButton } from '../misc/Add'
import { CustomAddSelect } from '../misc/CustomAddSelect'
import { CustomAddDatePicker } from '../misc/CustomAddDatePicker'
import { ReactComponent as NotesIcon } from '../../assets/icons/notes.svg'
import { tasksCategories } from '../../assets/fakeData'
import { TaskCategory } from '../../utils/ModuleTypes'

const AddTask: React.FC = () => {
  const [title, setTitle] = useState('')
  const [category, setCategory] = useState('0')
  const [dueDate, setDueDate] = useState(new Date())

  const categoryOptions = [
    { val: '0', label: 'Inbox' },
    ...(tasksCategories as TaskCategory[]).map(cat => ({
      val: cat.id,
      label: cat.name
    }))
  ]

  const handleCategoryChange = (e: React.ChangeEvent<HTMLSelectElement>) => {
    setCategory(e.target.value)
  }

  const handleSubmit = () => {
    console.log(title)
    console.log(category)
    console.log(dueDate)
  }

  return (
    <>
      <Styled.AddInput
        type="text"
        placeholder="Ex: Take out the trash"
        value={title}
        onChange={e => setTitle(e.target.value)}
      />

      <Styled.AddWidgetsContainer>
        <Styled.AddWidget>
          <DatePickerInput
            date={dueDate}
            setDate={setDueDate}
            customInput={<CustomAddDatePicker />}
          />
        </Styled.AddWidget>

        <Styled.AddWidget>
          <CustomAddSelect
            id="add-category"
            value={category}
            onChange={handleCategoryChange}
            options={categoryOptions}
            icon={<NotesIcon />}
          />
        </Styled.AddWidget>

        <AddSubmitButton handleSubmit={handleSubmit} />
      </Styled.AddWidgetsContainer>
    </>
  )
}

export default AddTask
