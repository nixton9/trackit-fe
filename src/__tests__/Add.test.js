import React from 'react'
import { render } from '@testing-library/react'
import Add from '../components/misc/Add'
import AddTask from '../components/tasks/AddTask'
import AddExpense from '../components/expenses/AddExpense'
import AddHabit from '../components/habits/AddHabit'
import { format } from 'date-fns'

describe('Add', () => {
  it('matches snapshot', () => {
    const { container } = render(<Add />)

    expect(container).toMatchSnapshot()
  })

  it('renders all the buttons and the title', () => {
    const { getByText } = render(<Add />)

    expect(getByText('What do you want to track?')).toBeTruthy()
    expect(getByText('Create a note')).toBeInTheDocument()
    expect(getByText('Create a task')).toBeInTheDocument()
    expect(getByText('Add an expense')).toBeInTheDocument()
    expect(getByText('Add an habit')).toBeInTheDocument()
  })
})

describe('AddTask', () => {
  it('matches snapshot', () => {
    const { container } = render(<AddTask />)

    expect(container).toBeInTheDocument()
  })

  it('renders the inputs', () => {
    const { getByPlaceholderText, getByDisplayValue, getByText } = render(
      <AddTask />
    )

    expect(getByPlaceholderText('Ex: Take out the trash')).toBeInTheDocument()
    expect(getByDisplayValue(format(new Date(), 'd MMM'))).toBeInTheDocument()
    expect(getByText('Inbox')).toBeInTheDocument()
  })
})

describe('AddExpense', () => {
  it('matches snapshot', () => {
    const { container } = render(<AddExpense />)

    expect(container).toBeInTheDocument()
  })

  it('renders the inputs', () => {
    const { getByPlaceholderText, getByDisplayValue, getByText } = render(
      <AddExpense />
    )

    expect(getByPlaceholderText('9.99')).toBeInTheDocument()
    expect(getByPlaceholderText('Ex: Dinner at mcdonalds')).toBeInTheDocument()
    expect(getByDisplayValue(format(new Date(), 'd MMM'))).toBeInTheDocument()
    expect(getByText('All')).toBeInTheDocument()
  })
})

describe('AddHabit', () => {
  it('matches snapshot', () => {
    const { container } = render(<AddHabit />)

    expect(container).toBeInTheDocument()
  })

  it('renders the inputs', () => {
    const { getByPlaceholderText } = render(<AddHabit />)

    expect(getByPlaceholderText('Ex: Eat healthy')).toBeInTheDocument()
  })
})
