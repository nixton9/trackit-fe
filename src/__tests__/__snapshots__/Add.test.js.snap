// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Add matches snapshot 1`] = `
<div>
  <div
    class="sc-AxgMl fBmUve"
  >
    <svg>
      plus.svg
    </svg>
  </div>
  <div
    class="sc-AxjAm khxSQh"
  >
    <h2
      class="sc-AxiKw jzoYkp"
    >
      What do you want to track?
    </h2>
    <div
      class="sc-AxhCb bCDkJU"
    >
      <button
        class="sc-AxheI gPascW"
      >
        <span>
          Create a note
        </span>
        <svg>
          notes.svg
        </svg>
      </button>
      <button
        class="sc-AxheI gPascW"
      >
        <span>
          Create a task
        </span>
        <svg>
          tasks.svg
        </svg>
      </button>
      <button
        class="sc-AxheI gPascW"
      >
        <span>
          Add an expense
        </span>
        <svg>
          expenses.svg
        </svg>
      </button>
      <button
        class="sc-AxheI gPascW"
      >
        <span>
          Add an habit
        </span>
        <svg>
          habits.svg
        </svg>
      </button>
    </div>
  </div>
  <div
    class="sc-AxirZ gpexXZ"
  />
</div>
`;

exports[`Add renders all the buttons and the title 1`] = `
<h2
  class="sc-AxiKw jzoYkp"
>
  What do you want to track?
</h2>
`;

exports[`Add renders all the buttons and the title 2`] = `
<span>
  Create a note
</span>
`;

exports[`Add renders all the buttons and the title 3`] = `
<span>
  Create a task
</span>
`;

exports[`Add renders all the buttons and the title 4`] = `
<span>
  Add an expense
</span>
`;

exports[`Add renders all the buttons and the title 5`] = `
<span>
  Add an habit
</span>
`;

exports[`AddTask matches snapshot 1`] = `
<div>
  <input
    class="sc-Axmtr doLFLl"
    placeholder="Ex: Take out the trash"
    type="text"
    value=""
  />
  <div
    class="sc-fzozJi iNhwPK"
  >
    <div
      class="sc-fzoLsD cnPPqL"
    >
      <div
        class="sc-AxhUy jCZCjm"
      >
        <div
          class="react-datepicker-wrapper"
        >
          <div
            class="react-datepicker__input-container"
          >
            <svg>
              calendr.svg
            </svg>
            <input
              type="text"
              value="3 Sep"
            />
          </div>
        </div>
      </div>
    </div>
    <div
      class="sc-fzoLsD cnPPqL"
    >
      <span>
        <svg>
          notes.svg
        </svg>
      </span>
      <div
        class="MuiInputBase-root MuiInput-root"
      >
        <div
          aria-haspopup="listbox"
          aria-labelledby="add-category"
          class="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input"
          id="add-category"
          role="button"
          tabindex="0"
        >
          Inbox
        </div>
        <input
          aria-hidden="true"
          class="MuiSelect-nativeInput"
          tabindex="-1"
          value="0"
        />
        <svg
          class="MuiSelect-icon"
        >
          chevron.svg
        </svg>
      </div>
    </div>
    <button
      class="sc-fzpans hckgSo"
    >
      <svg>
        chevron.svg
      </svg>
    </button>
  </div>
</div>
`;
